File: Iterator.py
"""
Iterator
Provide a way to access the elements of an aggregate object
sequentially without exposing its underlying representation.
"""

from abc import ABC, abstractmethod


class Iterator(ABC):
    @abstractmethod
    def has_next(self) -> bool:
        pass
    
    @abstractmethod
    def next(self):
        pass

class ConcreteIterator(Iterator):
    def __init__(self, items):
        self.items = items
        self.index = 0
    
    def has_next(self):
        return self.index < len(self.items)
    
    def next(self):
        item = self.items[self.index]
        self.index += 1
        return item

class Aggregate:
    def __init__(self):
        self.items = []
    
    def add(self, value):
        self.items.append(value)
    
    def get_items(self):
        return self.items


if __name__ == "__main__":
    a = Aggregate()
    a.add(1)
    a.add(2)
    a.add(3)
    
    i = ConcreteIterator(a.get_items())
    while i.has_next():
        print(i.next())

File: Iterator.java
/*****
 * Iterator
 * Provide a way to access the elements of an aggregate object
 * sequentially without exposing its underlying representation.
*****/

import java.util.ArrayList;
import java.util.List;


interface IteratorInterface<T> {
    boolean hasNext();
    T next();
}

class ConcreteIterator<T> implements IteratorInterface<T> {
    private final List<T> items;
    private int index = 0;
    
    public ConcreteIterator(List<T> items) { this.items = items; }
    
    @Override
    public boolean hasNext() { return index < items.size(); }
    
    @Override
    public T next() { return items.get(index++); }
}

class Aggregate<T> {
    private final List<T> items = new ArrayList<>();
    
    public void add(T value) { items.add(value); }
    public List<T> getItems() { return items; }
}


public class Iterator {
    public static void main(String[] args) {
        Aggregate<Integer> a = new Aggregate<>();
        a.add(1);
        a.add(2);
        a.add(3);
        
        IteratorInterface<Integer> i = new ConcreteIterator<>(a.getItems());
        while (i.hasNext()) System.out.println(i.next());
    }
}
